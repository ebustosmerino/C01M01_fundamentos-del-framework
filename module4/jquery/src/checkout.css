.checkout {
  display: flex;
}
.checkout__resume, .checkout__confirmation {
  padding: 20px;
}
.checkout__resume button {
  cursor: pointer;
}
.checkout__confirmation {
  display: none;
}
.checkout__confirmation select{
  margin-bottom: 20px;
}
.checkout__confirmation--active {
  display: block;
}
.checkout__confirmation--active fieldset{
  display: flex;
}
.checkout__confirmation .confirmation__instructor-result {
  display: none;
}
.checkout__confirmation .confirmation__instructor-result--active {
  display: block;
}
