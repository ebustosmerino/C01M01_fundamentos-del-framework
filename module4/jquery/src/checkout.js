// Código jQuery refactorizado para hacerlo testeable
const checkoutModule = {
	init: () => {
		//Page states and selectors
		checkoutModule.steps = {
			resume: {
				selector: '.checkout__resume',
				buttonNextStepSelector: '.checkout__resume button[type="button"]'
			},
			confirmation: {
				selector: '.checkout__confirmation',
				formSelector: '.checkout__confirmation form',
				instructorSelectionSelector: '.checkout__confirmation .confirmation__instructor-selection',
				instructorCourseSelector: '.checkout__confirmation .confirmation__instructor-course-selection',
				confirmationResultSelector: '.checkout__confirmation .confirmation__instructor-result',
			},
		};

		//Page listeners
		$(checkoutModule.steps.resume.buttonNextStepSelector)
			.on('click', checkoutModule.onClickResume);
	},
	onClickResume: ($event) => {
		const confirmationSelector = checkoutModule.steps.confirmation.selector;
		const formSelector = checkoutModule.steps.confirmation.formSelector;

		$(confirmationSelector).addClass('checkout__confirmation--active');
		$(formSelector)
			.on('submit', checkoutModule.onSubmit);
	},

	onSubmit: ($event) => {
		const {
			instructorSelectionSelector,
			instructorCourseSelector,
			confirmationResultSelector,
		} = checkoutModule.steps.confirmation;

		$.ajax({
			url: 'https://my-endpoint',
			type: 'POST',
			data: {
				instructor: $(instructorSelectionSelector).val(),
				course: $(instructorCourseSelector).val(),
			},
			success: () => {
				$(confirmationResultSelector)
					.text('datos enviados correctamente')
					.addClass('confirmation__instructor-result--active');
			},
			error: () => {
				$(confirmationResultSelector)
					.text('datos no pudieron ser enviados')
					.addClass('confirmation__instructor-result--active');
			}
		});

		$event.preventDefault();
		return false;
	},
};
$( document ).ready( checkoutModule.init );

window.checkoutModule = checkoutModule;

// Código jQuery sin estructurar
// $(document).ready(function () {
//   $('.checkout__resume button[type="button"]').on("click", function () {
//     $(".checkout__confirmation").addClass("checkout__confirmation--active");
//   });
//   $(".checkout__confirmation form").on("submit", function () {
//     $.ajax({
//       url: "https://my-endpoint",
//       type: "POST",
//       data: {
//         instructor: $('.checkout__confirmation .confirmation__instructor-selection').val(),
//         course: $('.checkout__confirmation .confirmation__instructor-course-selection').val(),
//       },
//       success: () => {
//         $('.checkout__confirmation .confirmation__instructor-result')
//           .text("datos enviados correctamente")
//           .addClass("confirmation__instructor-result--active");
//       },
//       error: () => {
//         $('.checkout__confirmation .confirmation__instructor-result')
//           .text("datos no pudieron ser enviados")
//           .addClass("confirmation__instructor-result--active");
//       },
//     });

//     $event.preventDefault();
//     return false;
//   });
// });