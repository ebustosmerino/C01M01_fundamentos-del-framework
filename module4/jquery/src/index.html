<!doctype html>
 <html lang="es">
  <head>
    <meta charset="utf-8">
    <title>Boolean Chile</title>
    <base href="/">
    <meta name="description" content="Boolean Chile: Cursos de habilidades digitales para todos.">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="icon" href="https://cdn.statically.io/gh/booleancl/activities-assets/954770f1/src/img/favicon.ico">
    <link rel="stylesheet" href="https://cdn.statically.io/gh/booleancl/activities-assets/954770f1/src/css/style.min.css">

    <!-- Css example for exercise  -->
    <link rel="stylesheet" href="checkout.css">
  </head>
  <body class="brand__orange">
    <header>
      <div class="header__top">
        <div class="brand__logo"></div>
      </div>
      <div class="header__bottom">
        <h1 class="title">Pruebas de Software para Frontend</h1>
        <h2 class="subtitle">Módulo 4 “Pruebas de Integración”</h2>
      </div>
    </header>
    <main class="content">
      <section class="checkout">
        <div class="checkout__resume">
          <h2>Resumen de tu compra</h2>
          <div>
            <h3>Pruebas de Integración en Frontend</h3>
            <div>
              <h4>Plan Básico</h4>
              <ul>
                <li> 6 clases de 1 hr</li>
                <li> Videollamada 1:1</li>
                <li> Chat personalizado para preguntas</li>
                <li> 1 Profesor dedicado experto en el tópico del programa</li>
                <li> Material Audiovisual y en formato PDF</li>
                <li> Evaluación e Informe de resultado</li>
              </ul>
              <button type="button" data-course-plan="custom">Reservar Hora</button>
            </div>
          </div>
        </div>
        <div class="checkout__confirmation">
          <h2>Selecciona la hora disponible</h2>
          <div>
            <form>
              <fieldset>
                <label for="instructor-selection">Elige a tu profesor</label>
                <select name="" id="instructor-selection" class="confirmation__instructor-selection">
                  <option>Selecciona un profesor</option>
                  <option value="gonzalo">Gonzalo Pincheira</option>
                  <option value="camilo">Camilo Flores</option>
                </select>
                <label for="instructor-course">Selecciona un curso</label>
                <select name="" id="instructor-course" class="confirmation__instructor-course-selection">
                  <option>Selecciona un curso</option>
                  <option value="angular">Angular</option>
                  <option value="vue">Vue</option>
                </select>
                <button type="submit">Enviar Solicitud</button>
              </fieldset>
            </form>
            <div class="confirmation__instructor-result"></div>
          </div>
        </div>
      </section>	
    </main>

    <script src="https://cdn.statically.io/gh/booleancl/activities-assets/f4acb9e9/src/lib/jquery.min.js"></script>
    <script src="checkout.js"></script>
  </body>
</html>
