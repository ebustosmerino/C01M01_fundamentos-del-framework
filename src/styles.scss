@import url(http://fonts.googleapis.com/css?family=Poppins:300italic,400italic,700italic,400,300,700);

* {
  font-family: 'Poppins', 'Times';
}

$tokens-color-white: rgb(255, 255, 255);
$tokens-color-almost-dark-blue: rgb(68, 72, 87);
$tokens-color-new: rgb(80, 85, 106);
$tokens-color-footer-background: rgb(9, 12, 22);
$tokens-color-course-planification: rgb(198, 201, 212);
$tokens-color-checkout-plan-default: rgb(198, 201, 212);

$tokens-color-grey-dark: rgb(34,36,68);
$tokens-color-grey-normal: rgb(69,71,101);
$tokens-color-grey-medium: rgb(90,95,117);
$tokens-color-grey-soft: rgb(138,141,158);
$tokens-color-grey-light: rgb(229,232,238);
$tokens-color-blue-dark: rgb(19,33,147);
$tokens-color-blue: rgb(6,64,246);
$tokens-color-blue-soft: rgb(41,135,255);
$tokens-color-blue-light: rgb(95,221,255);
$tokens-color-blue-bright: rgb(199,240,255);
$tokens-color-purple-dark: rgb(117,67,255);
$tokens-color-purple: rgb(137,81,255);
$tokens-color-purple-soft: rgb(180,133,255);
$tokens-color-pink: rgb(225,161,255);
$tokens-color-pink-dark: rgb(177,67,255);
$tokens-color-red: #ff7485;
$tokens-color-orange: rgb(252,186,158);
$tokens-color-yellow: rgb(255,245,169);
