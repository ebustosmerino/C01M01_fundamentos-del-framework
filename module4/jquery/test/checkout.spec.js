/*
	Necesario para hacer global jquery en Jest
*/
window.$ = require('../src/assets/lib/jquery.min');

/* 
	Incluimos el modulo. es necesario que el modulo sea
	asociado a la varible global window
*/
require('../src/checkout');

/* 
	Un template que representa el estado del DOM respecto
	del módulo con su estado inicial.
	Es posible utilizar algún loader si se tienen plantillas 
	con algun otro formato como pug, haml o html de manera
	de no escribir el template inline en este archivo

	Los archivos HTML pueden ser incluidos a través de "require"
	gracias a la configuración del archivo jest.config en la sección "transform"
*/
const template = require('./templates/checkout.spec.html');

describe('Checkout page', () => {
	/*
		Función que se ejecutará una única vez.
		En este caso la utilizamos para reemplazar el comportamiento original de la función
		$.ajax
		Si no se establece una implementación a través del método mockImplementation
		por defecto se asigna una función no operacional. (una función que no recibe argumentos
		que siempre retorna el valor undefined)
	*/
	beforeAll(() => {
		jest.spyOn(window.$, 'ajax');
	});

	beforeEach(() => {
		/*
			Por cada test asignamos el template con su estado inicial
			para volver determinista cada prueba de software y ejecutamos la función
			init para asociar los listeners nuevamente
		*/
		document.body.innerHTML = template;
		checkoutModule.init();
		
		window.$.ajax.mockReset();
	});

	/*
		Utilizamos Snapshot testing para hacer pruebas de software implícitas. Por lo tanto
		no vamos a hacer asersiones basadas en preguntar si ciertas clases fueron aplicadas
		sobre ciertos elementos. Vamos a sacar "una foto" de como quedó el DOM luego de aplicar
		las acciones y vamos a comparar con esta copia en las futuras ejecucioes de estas pruebas.
		En caso de que un cambio sea legítimo ya sea una corrección o modificación del algún detalle
		debemos actualizar los snapshots
	*/
	it('should active confirmation section when next Step button is clicked ', () => {
		expect(
			$(checkoutModule.steps.confirmation.selector).hasClass('checkout__confirmation--active')
		).not.toBe(true);

		$(checkoutModule.steps.resume.buttonNextStepSelector).click();

		// expect(
		// 	$(checkoutModule.steps.confirmation.selector).hasClass('checkout__confirmation--active')
		// ).toBe(true);
		expect(document.body.innerHTML).toMatchSnapshot();
	});

	it('should show success message if ajax call were ok', () => {
		$(checkoutModule.steps.resume.buttonNextStepSelector).click();
		window.$.ajax
			.mockImplementation((config) => config.success());
		$(checkoutModule.steps.confirmation.instructorSelectionSelector)
			.val('gonzalo');
		$(checkoutModule.steps.confirmation.instructorCourseSelector)
			.val('angular');
		
		$(checkoutModule.steps.confirmation.formSelector).submit();

		const ajaxCall = window.$.ajax.mock.calls[0][0];
		expect(ajaxCall.data).toEqual({
			instructor: 'gonzalo',
			course: 'angular',
		});
		expect(ajaxCall.url).toEqual('https://my-endpoint');
		expect(ajaxCall.type).toEqual('POST');
		expect(document.body.innerHTML).toMatchSnapshot();
	});

	it('should show error message if ajax call were ok', () => {
		$(checkoutModule.steps.resume.buttonNextStepSelector).click();
		window.$.ajax
			.mockImplementation((config) => config.error());		

		$(checkoutModule.steps.confirmation.formSelector).submit();

		const ajaxCall = window.$.ajax.mock.calls[0][0];
		expect(ajaxCall.data).toEqual({
			instructor: 'Selecciona un profesor',
			course: 'Selecciona un curso',
		});
		expect(ajaxCall.url).toEqual('https://my-endpoint');
		expect(ajaxCall.type).toEqual('POST');
		expect(document.body.innerHTML).toMatchSnapshot();
	});
});
