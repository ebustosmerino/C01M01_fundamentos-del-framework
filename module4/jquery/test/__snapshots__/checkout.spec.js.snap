// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkout page should active confirmation section when next Step button is clicked  1`] = `
"<section class=\\"checkout\\">
	<div class=\\"checkout__resume\\">
		<h2>Resumen de tu compra</h2>
		<div>
			<h3>Pruebas de Integración en Frontend</h3>
			<div>
				<h4>Plan Básico</h4>
				<ul>
					<li> 6 clases de 1 hr</li>
					<li> Videollamada 1:1</li>
					<li> Chat personalizado para preguntas</li>
					<li> 1 Profesor dedicado experto en el tópico del programa</li>
					<li> Material Audiovisual y en formato PDF</li>
					<li> Evaluación e Informe de resultado</li>
				</ul>
				<button type=\\"button\\" data-course-plan=\\"custom\\">Reservar Hora</button>
			</div>
		</div>
	</div>
	<div class=\\"checkout__confirmation checkout__confirmation--active\\">
		<h2>Selecciona la hora disponible</h2>
		<div>
			<form>
				<fieldset>
					<label for=\\"instructor-selection\\">Elige a tu profesor</label>
					<select name=\\"\\" id=\\"instructor-selection\\" class=\\"confirmation__instructor-selection\\">
						<option>Selecciona un profesor</option>
						<option value=\\"gonzalo\\">Gonzalo Pincheira</option>
						<option value=\\"camilo\\">Camilo Flores</option>
					</select>
					<label for=\\"instructor-course\\">Selecciona un curso</label>
					<select name=\\"\\" id=\\"instructor-course\\" class=\\"confirmation__instructor-course-selection\\">
						<option>Selecciona un curso</option>
						<option value=\\"angular\\">Angular</option>
						<option value=\\"vue\\">Vue</option>
					</select>
					<button type=\\"submit\\">Enviar Solicitud</button>
				</fieldset>
			</form>
			<div class=\\"confirmation__instructor-result\\"></div>
		</div>
	</div>
</section>	"
`;

exports[`Checkout page should show error message if ajax call were ok 1`] = `
"<section class=\\"checkout\\">
	<div class=\\"checkout__resume\\">
		<h2>Resumen de tu compra</h2>
		<div>
			<h3>Pruebas de Integración en Frontend</h3>
			<div>
				<h4>Plan Básico</h4>
				<ul>
					<li> 6 clases de 1 hr</li>
					<li> Videollamada 1:1</li>
					<li> Chat personalizado para preguntas</li>
					<li> 1 Profesor dedicado experto en el tópico del programa</li>
					<li> Material Audiovisual y en formato PDF</li>
					<li> Evaluación e Informe de resultado</li>
				</ul>
				<button type=\\"button\\" data-course-plan=\\"custom\\">Reservar Hora</button>
			</div>
		</div>
	</div>
	<div class=\\"checkout__confirmation checkout__confirmation--active\\">
		<h2>Selecciona la hora disponible</h2>
		<div>
			<form>
				<fieldset>
					<label for=\\"instructor-selection\\">Elige a tu profesor</label>
					<select name=\\"\\" id=\\"instructor-selection\\" class=\\"confirmation__instructor-selection\\">
						<option>Selecciona un profesor</option>
						<option value=\\"gonzalo\\">Gonzalo Pincheira</option>
						<option value=\\"camilo\\">Camilo Flores</option>
					</select>
					<label for=\\"instructor-course\\">Selecciona un curso</label>
					<select name=\\"\\" id=\\"instructor-course\\" class=\\"confirmation__instructor-course-selection\\">
						<option>Selecciona un curso</option>
						<option value=\\"angular\\">Angular</option>
						<option value=\\"vue\\">Vue</option>
					</select>
					<button type=\\"submit\\">Enviar Solicitud</button>
				</fieldset>
			</form>
			<div class=\\"confirmation__instructor-result confirmation__instructor-result--active\\">datos no pudieron ser enviados</div>
		</div>
	</div>
</section>	"
`;

exports[`Checkout page should show success message if ajax call were ok 1`] = `
"<section class=\\"checkout\\">
	<div class=\\"checkout__resume\\">
		<h2>Resumen de tu compra</h2>
		<div>
			<h3>Pruebas de Integración en Frontend</h3>
			<div>
				<h4>Plan Básico</h4>
				<ul>
					<li> 6 clases de 1 hr</li>
					<li> Videollamada 1:1</li>
					<li> Chat personalizado para preguntas</li>
					<li> 1 Profesor dedicado experto en el tópico del programa</li>
					<li> Material Audiovisual y en formato PDF</li>
					<li> Evaluación e Informe de resultado</li>
				</ul>
				<button type=\\"button\\" data-course-plan=\\"custom\\">Reservar Hora</button>
			</div>
		</div>
	</div>
	<div class=\\"checkout__confirmation checkout__confirmation--active\\">
		<h2>Selecciona la hora disponible</h2>
		<div>
			<form>
				<fieldset>
					<label for=\\"instructor-selection\\">Elige a tu profesor</label>
					<select name=\\"\\" id=\\"instructor-selection\\" class=\\"confirmation__instructor-selection\\">
						<option>Selecciona un profesor</option>
						<option value=\\"gonzalo\\">Gonzalo Pincheira</option>
						<option value=\\"camilo\\">Camilo Flores</option>
					</select>
					<label for=\\"instructor-course\\">Selecciona un curso</label>
					<select name=\\"\\" id=\\"instructor-course\\" class=\\"confirmation__instructor-course-selection\\">
						<option>Selecciona un curso</option>
						<option value=\\"angular\\">Angular</option>
						<option value=\\"vue\\">Vue</option>
					</select>
					<button type=\\"submit\\">Enviar Solicitud</button>
				</fieldset>
			</form>
			<div class=\\"confirmation__instructor-result confirmation__instructor-result--active\\">datos enviados correctamente</div>
		</div>
	</div>
</section>	"
`;
